### PR TITLE
sql: fixes database list with special characters

### PR DESCRIPTION
**Pull request by j82w**
The database name was not properly escaped which causes the
query to fail if the database name has a special character.
The endpoint still does not support names with `/`. This will
not be an issue since all ui is being converted to sql-over-http.

Example uri: `http://localhost:8080/_admin/v1/databases/my%20testdb`

part of: #94328

Release note (sql change): fixes databases list api when database name has special characters.

### CODE REVIEW

**Review by maryliag — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95209)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @j82w and @knz)

___
*[`pkg/server/admin_test.go` line 353 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95209#-NLfqGrjDTtubqQghEpr:-NLfqGrjDTtubqQghEps:b-9ycfzf) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bf29eadc0095403268d4d957ea49d9886042b8a/pkg/server/admin_test.go#L353)):*
> ```Go
> 		"test",
> 	)
> 	return ng.GenerateOne(42)
> ```

any way to guarantee this will have special characters?

**Comment by j82w** [comment]
*[`pkg/server/admin_test.go` line 353 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95209#-NLfqGrjDTtubqQghEpr:-NLga8TQ0-pboxqEhR1x:b-r7epeh) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bf29eadc0095403268d4d957ea49d9886042b8a/pkg/server/admin_test.go#L353)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

any way to guarantee this will have special characters?
</blockquote></details>

I updated the config so it will always have special characters.

**Review by maryliag — APPROVED** [review]
:lgtm:

> Reviewed 1 of 2 files at r1, 1 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95209#-:-NLgbOdd0vMq4akLTVCe:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @knz)

**Review by knz — COMMENTED** [review]
Nice.

> Reviewed 2 of 2 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95209#-:-NLgb7oQ3AoXwKW62bgY:b-83sjo4)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @j82w)

___
*[`pkg/server/index_usage_stats.go` line 391 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95209#-NLgaMn6AMTJtbEJk7UG:-NLgaMn71LRTO8-KwvOt:bwtsctz) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bf29eadc0095403268d4d957ea49d9886042b8a/pkg/server/index_usage_stats.go#L391)):*
> ```Go
> 	}
> 
> 	escDBName := tree.NameStringP(&dbName)
> ```

`tree.NameString(dbName)`
___
*[`pkg/server/admin_test.go` line 351 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95209#-NLg_gP_AX4411oydkSm:-NLg_gP_AX4411oydkSn:b1ufj9h) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bf29eadc0095403268d4d957ea49d9886042b8a/pkg/server/admin_test.go#L351)):*
> ```Go
> 		&cfg,
> 		rand,
> 		"test",
> ```

despite marylia's question, I think you should use `"a/b%s-c.d"` as a starting point for the generation.
___
*[`pkg/server/admin_test.go` line 353 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95209#-NLfqGrjDTtubqQghEpr:-NLg_Vev46eseLAvid4R:b-ds5fq) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bf29eadc0095403268d4d957ea49d9886042b8a/pkg/server/admin_test.go#L353)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

any way to guarantee this will have special characters?
</blockquote></details>

Yes when running stress tests, probabilistically it will contain special characters in 10% of cases.
___
*[`pkg/server/admin_test.go` line 372 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95209#-NLga8UC01ZHUZNKrekU:-NLga8UC01ZHUZNKrekV:b-3dofzi) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bf29eadc0095403268d4d957ea49d9886042b8a/pkg/server/admin_test.go#L372)):*
> ```Go
> 
> 	testDbName := generateRandomName()
> 	testDbEscaped := tree.NameStringP(&testDbName)
> ```

`tree.NameString(testDbName)` to avoid a heap allocation.
___
*[`pkg/server/admin_test.go` line 349 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95209#-NLga_iJ636YTMhpJYYJ:-NLga_iK1uuU1GERuanc:btqyso5) ([raw file](https://github.com/cockroachdb/cockroach/blob/f74273907bbaafd9c7b2944f4683114715f56750/pkg/server/admin_test.go#L349)):*
> ```Go
> 	// Force the name to always have special
> 	// characters.
> 	cfg := randident.NameGeneratorConfig{
> ```

No need for this if you take my previous suggestion.

**Comment by j82w** [comment]
*[`pkg/server/admin_test.go` line 351 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95209#-NLg_gP_AX4411oydkSm:-NLge28yDWzyhrsjmNMP:b11saln) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bf29eadc0095403268d4d957ea49d9886042b8a/pkg/server/admin_test.go#L351)):*
> a/b%s-c.d

As discussed the forward slash can not be supported with the current framework without exposing a new endpoint that takes in a hex encoded name to avoid the url escaping done by the language/framework. 

@maryliag do you know of any reasons why the rest endpoints (like:  `http://localhost:8080/_admin/v1/databases/rand%2Ftest?include_stats=true`) would need to support `/`? All the ui pages will support `/` when they are converted over to the sql-over-http.

**Review by j82w — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95209)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @knz and @maryliag)

___
*[`pkg/server/index_usage_stats.go` line 391 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95209#-NLgaMn6AMTJtbEJk7UG:-NLgccicBg4BcxzGMN2S:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bf29eadc0095403268d4d957ea49d9886042b8a/pkg/server/index_usage_stats.go#L391)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

`tree.NameString(dbName)`
</blockquote></details>

Done.
___
*[`pkg/server/admin_test.go` line 353 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95209#-NLfqGrjDTtubqQghEpr:-NM0JmHSAXuGKvJFBI4l:b9og75) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bf29eadc0095403268d4d957ea49d9886042b8a/pkg/server/admin_test.go#L353)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

Yes when running stress tests, probabilistically it will contain special characters in 10% of cases.
</blockquote></details>

This should be handled now that the input name has several special characters.
___
*[`pkg/server/admin_test.go` line 372 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95209#-NLga8UC01ZHUZNKrekU:-NLgcbft0_-m0Gx743rB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bf29eadc0095403268d4d957ea49d9886042b8a/pkg/server/admin_test.go#L372)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

`tree.NameString(testDbName)` to avoid a heap allocation.
</blockquote></details>

Done.
___
*[`pkg/server/admin_test.go` line 349 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95209#-NLga_iJ636YTMhpJYYJ:-NM0JjRLEJb7PUK6gfYe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f74273907bbaafd9c7b2944f4683114715f56750/pkg/server/admin_test.go#L349)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

No need for this if you take my previous suggestion.
</blockquote></details>

Done.

**Comment by knz** [comment]
Presumably we want a clearer error message in v22.2 when users encounter this situation, so an incremental improvement that we can backport is in order.

**Review by matthewtodd — CHANGES_REQUESTED** [review]
Nice! Just noticed a couple of print / log statements to remove.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95209#-:-NM0mqnyDcgBt2ZTZJnN:b-7w8j1q)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @j82w, @knz, and @maryliag)

___
*[`pkg/server/admin.go` line 385 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95209#-NM0mM-dDXDI3w49g5wP:-NM0mM-dDXDI3w49g5wQ:b-8gkfo1) ([raw file](https://github.com/cockroachdb/cockroach/blob/c560fc82ce94ac4abcba5a0b8aac265eb32ca486/pkg/server/admin.go#L385)):*
> ```Go
> 	limit, offset int,
> ) (_ *serverpb.DatabasesResponse, retErr error) {
> 	fmt.Println("databasesHelper")
> ```

Remove this before merging.
___
*[`pkg/server/admin.go` line 435 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95209#-NM0mOXb1so5foGhOEZW:-NM0mOXb1so5foGhOEZX:b-ax5s56) ([raw file](https://github.com/cockroachdb/cockroach/blob/c560fc82ce94ac4abcba5a0b8aac265eb32ca486/pkg/server/admin.go#L435)):*
> ```Go
> ) (_ *serverpb.DatabaseDetailsResponse, retErr error) {
> 	log.Infof(ctx, "DatabaseDetails: %s\n", req.Database)
> 	fmt.Printf("DatabaseDetails %s\n", req.Database)
> ```

Ditto these.

**Comment by j82w** [comment]
*[`pkg/server/admin_test.go` line 351 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95209#-NLg_gP_AX4411oydkSm:-NM47j6CAy5HdXPXi2x1:b-o0smgc) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bf29eadc0095403268d4d957ea49d9886042b8a/pkg/server/admin_test.go#L351)):*
Github comment from Raphael: 
> Presumably we want a clearer error message in v22.2 when users encounter this situation, so an incremental improvement that we can backport is in order.

@kevin-v-ngo can you confirm if any customers have a `/` in the name? As far as I know we don't currently have any customer reported issue regarding special characters in db or table names. I don't think it is worth the time and risk to backport a new api that will just be replaced with sql-over-http in 23.1. Also encoding/decoding all the names adds additional unnecessary overhead for most users.

**Comment by knz** [comment]
> can you confirm if any customers have a / in the name? As far as I know we don't currently have any customer reported issue regarding special characters in db or table names

The point i'm making is not that we should enable those hypothetical (and likely non-existing) customers.

My point is that we should ensure the engine reports a decent and clear error when that case is encountered, for those users who explore and run into the problem. Currently the fault is cryptic.

**Comment by knz** [comment]
one way to identify the error is to do the check in the front-end code, before the query to the back-end is formed.

**Comment by j82w** [comment]
> one way to identify the error is to do the check in the front-end code, before the query to the back-end is formed.

This PR is focusing only on the backend fix for the databases api. There will be a separate PR for all the UI fixes. I can include the check their that will get removed once the sql-over-http conversion is done. That way it will be included in the backport to v22.2

**Comment by knz** [comment]
>  I can include the check their that will get removed once the sql-over-http conversion is done. That way it will be included in the backport to v22.2

Yes that sounds good.

**Review by knz — APPROVED** [review]
LGTM modulo nit

> Reviewed 2 of 3 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95209#-:-NM4_dHz9hk4gE4dJd1M:b-my0qfe)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @j82w)

___
*[`pkg/server/admin_test.go` line 349 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95209#-NLga_iJ636YTMhpJYYJ:-NM4_8UO78Z9VxBrURjC:b1h48zg) ([raw file](https://github.com/cockroachdb/cockroach/blob/f74273907bbaafd9c7b2944f4683114715f56750/pkg/server/admin_test.go#L349)):*
<details><summary><i>Previously, j82w (Jake) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

1. don't forget to call .Validate on cfg.
2. maybe add `cfg.Punctuate = -1` before Validate so that the random generator won't introduce `/` in the generated name; and you can refer to the remaining issue in a TODO.

**Review by matthewtodd — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95209)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @j82w)

**Review by j82w — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95209)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @knz)

___
*[`pkg/server/admin_test.go` line 349 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95209#-NLga_iJ636YTMhpJYYJ:-NM4w3SvED04qhDrgZyA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f74273907bbaafd9c7b2944f4683114715f56750/pkg/server/admin_test.go#L349)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

1. don't forget to call .Validate on cfg.
2. maybe add `cfg.Punctuate = -1` before Validate so that the random generator won't introduce `/` in the generated name; and you can refer to the remaining issue in a TODO.
</blockquote></details>

Done.
___
*[`pkg/server/admin.go` line 385 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/95209#-NM0mM-dDXDI3w49g5wP:-NM3uaOMA3xuRS34gv62:b-edpef9) ([raw file](https://github.com/cockroachdb/cockroach/blob/c560fc82ce94ac4abcba5a0b8aac265eb32ca486/pkg/server/admin.go#L385)):*
<details><summary><i>Previously, matthewtodd (Matthew Todd) wrote…</i></summary><blockquote>

Remove this before merging.
</blockquote></details>

Thanks for catching this. Fixed in latest iteration.

**Comment by j82w** [comment]
blathers backport 22.2

**Pull request merged**
